[SPARK-5479] [yarn] Handle --py-files correctly in YARN.

vanzin opened the pull request:
The bug description is a little misleading: the actual issue is that
.py files are not handled correctly when distributed by YARN. They're
added to "spark.submit.pyFiles", which, when processed by context.py,
explicitly whitelists certain extensions (see PACKAGE_EXTENSIONS),
and that does not include .py files.

On top of that, archives were not handled at all! They made it to the
driver's python path, but never made it to executors, since the mechanism
used to propagate their location (spark.submit.pyFiles) only works on
the driver side.

So, instead, ignore "spark.submit.pyFiles" and just build PYTHONPATH
correctly for both driver and executors. Individual .py files are
placed in a subdirectory of the container's local dir in the cluster,
which is then added to the python path. Archives are added directly.

The change, as a side effect, ends up solving the symptom described
in the bug. The issue was not that the files were not being distributed,
but that they were never made visible to the python application
running under Spark.

Also included is a proper unit test for running python on YARN, which
broke in several different ways with the previous code.

A short walk around of the changes:
- SparkSubmit does not try to be smart about how YARN handles python
  files anymore. It just passes down the configs to the YARN client
  code.
- The YARN client distributes python files and archives differently,
  placing the files in a subdirectory.
- The YARN client now sets PYTHONPATH for the processes it launches;
  to properly handle different locations, it uses YARN's support for
  embedding env variables, so to avoid YARN expanding those at the
  wrong time, SparkConf is now propagated to the AM using a conf file
  instead of command line options.
- Because the Client initialization code is a maze of implicit
  dependencies, some code needed to be moved around to make sure
  all needed state was available when the code ran.
- The pyspark tests in YarnClusterSuite now actually distribute and try
  to use both a python file and an archive containing a different python
  module. Also added a yarn-client tests for completeness.
- I cleaned up some of the code around distributing files to YARN, to
  avoid adding more copied & pasted code to handle the new files being
  distributed.

Commit:
[SPARK-5479] [yarn] Handle --py-files correctly in YARN.

The bug description is a little misleading: the actual issue is that
.py files are not handled correctly when distributed by YARN. They're
added to "spark.submit.pyFiles", which, when processed by context.py,
explicitly whitelists certain extensions (see PACKAGE_EXTENSIONS),
and that does not include .py files.

On top of that, archives were not handled at all! They made it to the
driver's python path, but never made it to executors, since the mechanism
used to propagate their location (spark.submit.pyFiles) only works on
the driver side.

So, instead, ignore "spark.submit.pyFiles" and just build PYTHONPATH
correctly for both driver and executors. Individual .py files are
placed in a subdirectory of the container's local dir in the cluster,
which is then added to the python path. Archives are added directly.

The change, as a side effect, ends up solving the symptom described
in the bug. The issue was not that the files were not being distributed,
but that they were never made visible to the python application
running under Spark.

Also included is a proper unit test for running python on YARN, which
broke in several different ways with the previous code.

A short walk around of the changes:
- SparkSubmit does not try to be smart about how YARN handles python
  files anymore. It just passes down the configs to the YARN client
  code.
- The YARN client distributes python files and archives differently,
  placing the files in a subdirectory.
- The YARN client now sets PYTHONPATH for the processes it launches;
  to properly handle different locations, it uses YARN's support for
  embedding env variables, so to avoid YARN expanding those at the
  wrong time, SparkConf is now propagate to the AM using a conf file
  instead of command line options.
- Because the Client initialization code is a maze of implicit
  dependencies, some code needed to be moved around to make sure
  all needed state was available when the code ran.
- The pyspark tests in YarnClusterSuite now actually distribute and try
  to use both a python file and an archive containing a different python
  module. Also added a yarn-client tests for completeness.
- I cleaned up some of the code around distributing files to YARN, to
  avoid adding more copied & pasted code to handle the new files being
  distributed.

vanzin commented:
Tests run:
- unit tests against hadoop 2.5.0
- real yarn cluster, client and cluster modes (CDH 5.4). In cluster mode, I used a .py file stored in HDFS.
- standalone client mode
- local mode

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33352 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33352/consoleFull) for   PR 6360 at commit [`943cbf4`](https://github.com/apache/spark/commit/943cbf450d32f49a16091247f2bf7e0679d184ae).

SparkQA commented:
[Test build #33352 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33352/consoleFull) for   PR 6360 at commit [`943cbf4`](https://github.com/apache/spark/commit/943cbf450d32f49a16091247f2bf7e0679d184ae).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33352/
Test FAILed.

Commit:
Style.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33363 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33363/consoleFull) for   PR 6360 at commit [`09045f1`](https://github.com/apache/spark/commit/09045f1db1f51657a040fec7c27c0ed1ed1463cb).

SparkQA commented:
[Test build #33363 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33363/consoleFull) for   PR 6360 at commit [`09045f1`](https://github.com/apache/spark/commit/09045f1db1f51657a040fec7c27c0ed1ed1463cb).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33363/
Test FAILed.

vanzin commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33381 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33381/consoleFull) for   PR 6360 at commit [`09045f1`](https://github.com/apache/spark/commit/09045f1db1f51657a040fec7c27c0ed1ed1463cb).

SparkQA commented:
[Test build #33381 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33381/consoleFull) for   PR 6360 at commit [`09045f1`](https://github.com/apache/spark/commit/09045f1db1f51657a040fec7c27c0ed1ed1463cb).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33381/
Test PASSed.

lianhuiwang commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -372,11 +381,29 @@ private[spark] class Client(
     }
 
+    if (args.primaryPyFile != null) {
+      distribute(args.primaryPyFile)
```
i think primaryPyFile is only used in applicationMaster, so it is distributed to appMaster.

vanzin replied:
Don't you need the file around when you run code in the executors? How is the code shipped to executors otherwise?

I can try it out, but it feels weird to have used code only in the AM.

vanzin replied:
Hmm, strangely enough it seems things work without distributing the main file. Wonder if that's safe in all cases or just a side-effect of the really simple test code, though.

srowen commented:
I'm not an expert on this code by any means. The problem description and solution sound reasonable and match up with the change. I skimmed the change and didn't see anything I'd flag. Tests pass. I'd love to get more Python folks to look at it, like @davies (and I see @lianhuiwang is already looking too), but this appears to be a good cleanup and problem solving change for 1.5 / `master`.

Commit:
No need to distribute primary file to executors.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33481 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33481/consoleFull) for   PR 6360 at commit [`7fe3cd4`](https://github.com/apache/spark/commit/7fe3cd4aa603032ded2461339d6326a6c93caa78).

SparkQA commented:
[Test build #33481 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33481/consoleFull) for   PR 6360 at commit [`7fe3cd4`](https://github.com/apache/spark/commit/7fe3cd4aa603032ded2461339d6326a6c93caa78).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33481/
Test PASSed.

vanzin commented:
Ping @tgravescs @andrewor14

tgravescs commented:
@vanzin  sorry I'm going to be on vacation for the next week

Commit:
Merge branch 'master' into SPARK-5479

Conflicts:
	core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33921 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33921/consoleFull) for   PR 6360 at commit [`cdbb990`](https://github.com/apache/spark/commit/cdbb9903c13a7b18bb82fd1498650dba2da21f53).

SparkQA commented:
[Test build #33921 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33921/consoleFull) for   PR 6360 at commit [`cdbb990`](https://github.com/apache/spark/commit/cdbb9903c13a7b18bb82fd1498650dba2da21f53).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Scalastyle.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33995 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33995/consoleFull) for   PR 6360 at commit [`220358b`](https://github.com/apache/spark/commit/220358bc53ff2cc97c8f31d00a7166563f82046d).

SparkQA commented:
**[Test build #33995 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33995/consoleFull)**     for PR 6360 at commit [`220358b`](https://github.com/apache/spark/commit/220358bc53ff2cc97c8f31d00a7166563f82046d)     after a configured wait of `150m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

vanzin commented:
jenkins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34016 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34016/consoleFull) for   PR 6360 at commit [`220358b`](https://github.com/apache/spark/commit/220358bc53ff2cc97c8f31d00a7166563f82046d).

SparkQA commented:
[Test build #34016 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34016/consoleFull) for   PR 6360 at commit [`220358b`](https://github.com/apache/spark/commit/220358bc53ff2cc97c8f31d00a7166563f82046d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

srowen commented:
@tgravescs shall we give you a few more days to take a look? Tests pass and I tend to trust Mr Vanzin here. @davies do you have any input from a Python perspective?

andrewor14 commented:
Hey I'll take a look at this later today, sorry for slipping.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -315,62 +315,59 @@ object SparkSubmit {
-          val pyArchivesFile = new File(pyLibPath, "pyspark.zip")
-          if (!pyArchivesFile.exists()) {
-            printErrorAndExit("pyspark.zip does not exist for python application in yarn mode.")
+    if (args.isPython) {
```
I actually find the prior organization of this code a little easier to follow. It read
```
if (args.isPython && clusterManager == YARN) { /* do something */ }
if (args.isPython && deployMode == CLIENT) { /* do something else */ }
```
So if I only cared about python but not YARN I can just skip the first block and read the second block. Now it's one giant python block. Not a big deal, but if merging these isn't strictly necessary I would prefer the old structure.

andrewor14 replied:
(also it'll be easier to back port stuff)

vanzin replied:
Ok, I'll revert to the old style. I actually would like to move the YARN code to Client.scala but didn't want to pile that up onto this change.

andrewor14 replied:
yeah that might be a good change

vanzin replied:
I ended up doing that in this change, since I needed to modify SparkSubmit anyway because the two conditions needed to run in the opposite order... well, long story short, the YARN code is now in Client.scala, where it should have always been.

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -315,62 +315,59 @@ object SparkSubmit {
+            if (!pyArchivesFile.exists()) {
+              printErrorAndExit("pyspark.zip does not exist for python application in yarn mode.")
+            }
+            val py4jFile = new File(pyLibPath, "py4j-0.8.2.1-src.zip")
```
not your code, but at some point we should factor this out instead of hardcoding the version here

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -315,62 +315,59 @@ object SparkSubmit {
+          args.childArgs = ArrayBuffer(args.primaryResource, args.pyFiles) ++ args.childArgs
+          if (clusterManager != YARN) {
+            args.files = mergeFileLists(args.files, args.primaryResource)
+          }
```
can you add a comment to explain why we don't want to do this in YARN?

andrewor14 replied:
just so I didn't miss anything, the only functional changes in this block are that you added YARN checks here and in L364, right?

andrewor14 commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -387,13 +384,6 @@ object SparkSubmit {
-
       // In yarn-cluster mode for a R app, add primary resource to files
       // that can be distributed with the job
       if (args.isR) {
```
now that you removed the python stuff you can simplify this to `if (isYarnCluster && args.isR) {`

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala:
```diff
@@ -487,9 +495,9 @@ private[spark] class ApplicationMaster(
     if (args.primaryPyFile != null && args.primaryPyFile.endsWith(".py")) {
-      System.setProperty("spark.submit.pyFiles",
-        PythonRunner.formatPaths(args.pyFiles).mkString(","))
+      userArgs = Seq(args.primaryPyFile, "") ++ userArgs
```
why the empty string?

vanzin replied:
Because PythonRunner needs that argument to be present.

andrewor14 replied:
can you add comment

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -296,6 +284,53 @@ private[spark] class Client(
+        if (addDistributedUri(localURI)) {
+          val localPath = getQualifiedLocalPath(localURI, hadoopConf)
+          val linkname = targetDir.map(_ + "/").getOrElse("") +
+            destName.orElse(Option(localURI.getFragment())).getOrElse(localPath.getName())
```
can you explain why we use the fragment here? can you add a comment?

vanzin replied:
I don't really know. That was part of the previous code.

andrewor14 replied:
(when you come back from your vacation) @tgravescs do you know?

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -296,6 +284,53 @@ private[spark] class Client(
+     * @param targetDir Subdirectory where to place the file.
+     * @param appMasterOnly Whether to distribute only to the AM.
+     * @return A 2-tuple. First item is whether the file is a "local:" URI. Second item is the
+     *         localized path for non-local paths, or the input `path` for local paths.
```
The second item can be null. Can you document when this is the case to avoid NPEs in the future if this isn't used carefully?

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -296,6 +284,53 @@ private[spark] class Client(
+      val (_, localizedPath) = distribute(args.keytab,
+        destName = Some(sparkConf.get("spark.yarn.keytab")),
+        appMasterOnly = true)
+      require(localizedPath != null)
```
do we require here because the key tab should only be distributed once? If so can you add a comment about that here can maybe a fail message for the require?

vanzin replied:
I don't think it will ever happen since this is the first call to `distribute`, so it's mostly paranoia.

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -296,6 +284,53 @@ private[spark] class Client(
+        resType: LocalResourceType = LocalResourceType.FILE,
+        destName: Option[String] = None,
+        targetDir: Option[String] = None,
+        appMasterOnly: Boolean = false): (Boolean, String) = {
```
thanks for cleaning this up with a helpful java doc!

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -296,6 +284,53 @@ private[spark] class Client(
+     * @return A 2-tuple. First item is whether the file is a "local:" URI. Second item is the
+     *         localized path for non-local paths, or the input `path` for local paths.
+     */
+    def distribute(
```
can be private I think

vanzin replied:
Nested methods cannot have modifiers.

andrewor14 replied:
oh I didn't even realize this is in a method... this YARN code is monstrous

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -307,33 +342,18 @@ private[spark] class Client(
-          }
-        } else if (confKey != null) {
+    ).foreach { case (destName, path, confKey) =>
+      if (path != null && !path.isEmpty()) {
```
this is very slightly different from before; before we trimmed the path before checking whether it's empty. Maybe it's not necessary but just to make sure we're aware of this

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -307,33 +342,18 @@ private[spark] class Client(
           // If the resource is intended for local use only, handle this downstream
           // by setting the appropriate property
-          sparkConf.set(confKey, localPath)
+          require(localizedPath != null)
```
the comment above does not describe this require, so we should move it down 1 line (also a fail message here would be good)

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -349,21 +369,10 @@ private[spark] class Client(
-          } else if (addToClasspath) {
-            // Resource is intended for local use only and should be added to the class path
-            cachedSecondaryJarLinks += file.trim()
+          val (isLocal, localizedPath) = distribute(file, resType = resType)
```
isLocal is not used

andrewor14 replied:
actually, shouldn't it be used?
```
if (isLocal && addToClassPath) { ... }
```

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -296,6 +284,53 @@ private[spark] class Client(
+        targetDir: Option[String] = None,
+        appMasterOnly: Boolean = false): (Boolean, String) = {
+      val localURI = new URI(path.trim())
+      if (localURI.getScheme != LOCAL_SCHEME) {
```
not your code, but can you add a comment here to explain the differences between how we handle resources with local schemes and ones without

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -471,9 +505,6 @@ private[spark] class Client(
     }
-    // Set the environment variables to be passed on to the executors.
-    distCacheMgr.setDistFilesEnv(env)
-    distCacheMgr.setDistArchivesEnv(env)
```
I noticed that we order some of these code blocks around in this patch. Is there a particular reason for it? Have you verified that it's safe to do so?

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -899,8 +938,21 @@ object Client extends Logging {
+  // Subdirectory where the user's Spark and Hadoop config files will be placed.
+  val LOCALIZED_CONF_DIR = "__spark_conf__"
+
+  // Name fo the file in the conf archive containing Spark configuration.
```
Name of

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -899,8 +938,21 @@ object Client extends Logging {
+  // `SparkConf.setExecutorEnv`, because that would cause it to propagate to the python
+  // code via the configuration, and then override the process's own environment when
+  // launching workers. Since it contains variables that are expanded by YARN, that cannot
+  // happen.
```
this is a little long-winded. Did you mean to say that there may be variables in this python path that are substituted on the executor side?

vanzin replied:
I'll reword it.

vanzin replied:
Actually, re-tested without this and it seems like using the property file handles the original issue already. So I'll just remove it.

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -548,8 +596,13 @@ private[spark] class Client(
+
+    // Set the environment variables to be passed on to the executors.
+    distCacheMgr.setDistFilesEnv(launchEnv)
+    distCacheMgr.setDistArchivesEnv(launchEnv)
```
I noticed that we order some of these code blocks around in this patch. Is there a particular reason for it? Have you verified that it's safe to do so?

vanzin replied:
From the change description:

> Because the Client initialization code is a maze of implicit dependencies, some code needed to be moved around to make sure all needed state was available when the code ran.

So yes, I have verified it.

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -490,15 +521,32 @@ private[spark] class Client(
+    // If pyFiles contains any .py files, we need to add LOCALIZED_PYTHON_DIR to the PYTHONPATH
+    // of the container processes too. Add all non-.py files directly to PYTHONPATH.
+    //
+    // NOTE: the code currently does not handle .py files defined with a "local:" scheme.
```
should we add a warning when this happens?

vanzin replied:
This is a general problem with pyspark in general (not handling local: URIs), not particular to YARN...

andrewor14 commented:
@vanzin This looks great. I think the approach makes a lot of sense. The inline suggestions I have are mainly documentation and code organization.

A higher level comment I have is that there is currently no high-level documentation on how "Python is handled in YARN cluster mode", say vs client mode. To understand how it works the reader right now has to go through the code and look at how container envs are set up and read the comments there. I think it would be very useful to add a large block comment somewhere in `SparkSubmit` or `Client` to explain this on a higher level. Some things to potentially include in this block comment:
- pyspark / py4j files are automatically zipped up and put on the `PYTHONPATH` of all executors
- .py files are put in a special dir, which is then put on the `PYTHONPATH`
- archive python files (.egg, .zip etc.) are put directly on the `PYTHONPATH`
- primary python file is only distributed in cluster mode and only to the AM (because the driver runs there)
- where these python files are (both primary and secondary)

Another thing I was wondering is: was there a particular reason behind the change to how Spark properties are distributed (command line java opts -> `--properties-file`)? I think it's OK to keep it as is, but it would be good to put that in a java doc somewhere. (It's still a surprise to me how `Client` is a thousand line file but doesn't even have a high level java doc on the class)

TL;DR I think this is a good change for master. It's pretty close to being merged aside from a few documentation suggestions.

vanzin commented:
> A higher level comment I have is that there is currently no high-level documentation on how "Python is handled in YARN cluster mode", say vs client mode.

I think the main goal is to not have a difference between YARN client vs. cluster mode. This change goes a long way towards that goal.

vanzin commented:
> Another thing I was wondering is: was there a particular reason behind the change to how Spark properties are distributed (command line java opts -> --properties-file)?

Mainly because if you do it in the command line YARN will replace any env variable references in the values, which in this case we do not want (because we want the expansion to occur on the node where executors will be launched, no where the AM is run).

Commit:
Merge branch 'master' into SPARK-5479

Commit:
Review feedback.

Commit:
Move some code to the YARN module.

This gets rid of an ordering issue with the code in SparkSubmit that
was easy to miss.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34245 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34245/consoleFull) for   PR 6360 at commit [`705571d`](https://github.com/apache/spark/commit/705571d6f999f3a8a4681626e79b4bf1cd134eb1).

Commit:
Actually run pyspark in client mode.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34248 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34248/consoleFull) for   PR 6360 at commit [`c8e5a82`](https://github.com/apache/spark/commit/c8e5a82b9add370b0b03f0af98af7fc167ab22a5).

Commit:
Only pyspark cares about python archives.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34249 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34249/consoleFull) for   PR 6360 at commit [`c743778`](https://github.com/apache/spark/commit/c74377873bde48341e4eda9f33c7f79f6cf2cc74).

SparkQA commented:
[Test build #34249 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34249/consoleFull) for   PR 6360 at commit [`c743778`](https://github.com/apache/spark/commit/c74377873bde48341e4eda9f33c7f79f6cf2cc74).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #34245 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34245/consoleFull) for   PR 6360 at commit [`705571d`](https://github.com/apache/spark/commit/705571d6f999f3a8a4681626e79b4bf1cd134eb1).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
**[Test build #34248 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34248/consoleFull)**     for PR 6360 at commit [`c8e5a82`](https://github.com/apache/spark/commit/c8e5a82b9add370b0b03f0af98af7fc167ab22a5)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

vanzin commented:
Jekins, retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34286 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34286/consoleFull) for   PR 6360 at commit [`c743778`](https://github.com/apache/spark/commit/c74377873bde48341e4eda9f33c7f79f6cf2cc74).

SparkQA commented:
[Test build #34286 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34286/console) for   PR 6360 at commit [`c743778`](https://github.com/apache/spark/commit/c74377873bde48341e4eda9f33c7f79f6cf2cc74).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

Commit:
Merge branch 'master' into SPARK-5479

vanzin commented:
Hmm, moving the archive-finding code to Client.scala caused the failure... looking.

Commit:
Fix yarn-client mode.

Need a different way to tell Client.scala to distribute pyspark libs.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34316 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34316/consoleFull) for   PR 6360 at commit [`c47501f`](https://github.com/apache/spark/commit/c47501f650b39beeaf35829cd62fd5b2c6a7451e).

SparkQA commented:
[Test build #34316 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34316/console) for   PR 6360 at commit [`c47501f`](https://github.com/apache/spark/commit/c47501f650b39beeaf35829cd62fd5b2c6a7451e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

srowen commented:
@andrewor14 how's this looking to you? I think the review feedback was incorporated. Do we need @tgravescs to have a look?

andrewor14 commented:
I'll take a look again later today, but last time I checked this is already in a good state.

tgravescs commented:
no need to wait for me. I had taken a quick look and think the overall approach looks good.

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala:
```diff
@@ -503,8 +513,8 @@ private[spark] class ApplicationMaster(
-          val mainArgs = new Array[String](args.userArgs.size)
-          args.userArgs.copyToArray(mainArgs, 0, args.userArgs.size)
+          val mainArgs = new Array[String](userArgs.size)
+          userArgs.copyToArray(mainArgs, 0, userArgs.size)
```
can't this just be `val mainArgs = userArgs.toArray`? Am I missing something?

vanzin replied:
This comes from all the way back when there was `yarn/alpha` and `yarn/stable`, but yeah, it doesn't look necessary.

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala:
```diff
@@ -490,9 +498,11 @@ private[spark] class ApplicationMaster(
-      System.setProperty("spark.submit.pyFiles",
-        PythonRunner.formatPaths(args.pyFiles).mkString(","))
+      // Second argument is the list of files to add to PYTHONPATH, which Client.scala already
+      // handles, so it's empty.
```
It took a while for me to get this. This is actually referring to the arguments to `o.a.s.deploy.PythonRunner`'s main method. We need to document this because others who are not familiar with Spark submit at all will most certainly get lost around here.

andrewor14 commented on yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:
```diff
@@ -548,8 +606,19 @@ private[spark] class Client(
-    val localResources = prepareLocalResources(appStagingDir)
-    val launchEnv = setupLaunchEnv(appStagingDir)
+    val pySparkArchives =
+      if (sys.props.getOrElse("spark.yarn.isPython", null) == "true") {
```
could this be `sys.props.getOrElse("...", "false").toBoolean` so it handles cases as well?

andrewor14 commented:
Ok, I tested this on a Hadoop 2.4 cluster and verified that both client and cluster modes work with `--py-files`. I'll merge this once you address the last round of comments.

Commit:
Feedback.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34629 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34629/consoleFull) for   PR 6360 at commit [`bcaf7e6`](https://github.com/apache/spark/commit/bcaf7e66f62043286a4dde93a9d29b6e462e5804).

andrewor14 commented:
One higher level thing I'd like to mention is that the YARN code is getting quite difficult to review. We cleaned up a lot of the duplicate code and added documentation everywhere at one point, and that helped a lot. However, it is still difficult to follow the code or ensure its correctness because of a few things:

- Monstrous 100-200 lines methods like `setupLaunchEnv` and `prepareLocalResources`. These will only grow over time.
- Lack of high level documentation of what important classes like `Client`, `ApplicationMaster` and `ExecutorRunnable` do. Most of these don't even have class-level java docs.
- There are many implicit assumptions (e.g. use URI fragments if the local scheme is present) that are carried over from legacy code and undocumented
- Test coverage is poor. This is related to the first point; having large methods makes it difficult to write tests at fine granularity.

Rather than suggesting that we fix all these at once, I'm just jotting down my thoughts so we don't forget to fix these down the road. I'll add these thoughts to the umbrella YARN clean up JIRA.

vanzin commented:
> the YARN code is getting quite difficult to review

Yeah, that comes from a long time. It really needs some cleaning up. It's not just the size of methods, but the implicit dependencies they have on one another (method A sets some instance field that is needed by method B, so they must be executed in that order or things break). That's the trickiest thing when messing with this code (the size is not that big of an issue IMO).

andrewor14 commented:
OK, I've updated [SPARK-3492](https://issues.apache.org/jira/browse/SPARK-3492). Yes, all implicit dependencies should be documented or even removed.

SparkQA commented:
[Test build #34629 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34629/console) for   PR 6360 at commit [`bcaf7e6`](https://github.com/apache/spark/commit/bcaf7e66f62043286a4dde93a9d29b6e462e5804).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented:
LGTM merging into master thanks @vanzin.

zjffdu commented:
@vanzin  I am reading the yarn related code specially on org.apache.spark.deploy.yarn.Client.scala
Do you know where LOCAL_SCHEME("local") come from ? As I know we use file:// to represent a local resource, so not sure where "local" come from. 

Another question is that if I specify spark.yarn.jar as a hdfs location, the yarn client will still copy it to staging directory, I don't know why we do this. Would just use the hdfs file as LocalResource without copying much easier ?

vanzin commented:
@zjffdu please don't comment on closed PRs, especially ancient ones; that's super confusing. Use the dev@ list for that.

To answer your questions: the "local:" scheme is defined in Spark's documentation. Defining a "hdfs" location for `spark.yarn.jar` should not cause it to be copied to the staging dir, unless they are on different HDFS servers. Maybe the way you're defining the value is confusing the code in Client.scala that checks that.

zjffdu commented:
Thanks @vanzin, my fault, I specify the hdfs location using hostname, while it is ip address in core-site.xml (anyway, maybe we can improve it here)
